### PR TITLE
⚡️ Implement add_bridge

### PR DESCRIPTION
**Pull request by blegat**
~~Requires https://github.com/JuliaOpt/MathOptInterface.jl/pull/617~~

### CODE REVIEW

File: src/JuMP.jl [hunk 1]
```diff
@@ -165,6 +165,9 @@ mutable struct Model <: AbstractModel
     # In MANUAL and AUTOMATIC modes, CachingOptimizer.
     # In DIRECT mode, will hold an AbstractOptimizer.
     moi_backend::MOI.AbstractOptimizer
+    # List of additional bridges to add in addition to the ones added in
```
Comment (mlubin) [inline src/JuMP.jl]: "additional" is repetitive

File: src/JuMP.jl [hunk 2]
```diff
@@ -165,6 +165,9 @@ mutable struct Model <: AbstractModel
...
+    # `MOI.Bridges.fullbridgeoptimizer`.
```
Comment (mlubin) [inline src/JuMP.jl]: `full_bridge_optimizer`

File: src/JuMP.jl [hunk 3]
```diff
@@ -350,6 +354,39 @@ bridge_constraints(model::Model) = moi_bridge_constraints(backend(model))
...
+Add `BridgeType` to the list of bridges that can be used to transform
+unsupported constraints into an equivalent formulation using only constraints
+supported by the optimizer.
+"""
+function add_bridge(model::Model,
+                    BridgeType::Type{<:MOI.Bridges.AbstractBridge})
+    push!(model.bridge_types, BridgeType)
+    moi_add_bridge(JuMP.backend(model), BridgeType)
+end
+# The type of backend(model) is unknown so we directly redirect to another
```
Comment (mlubin) [inline src/JuMP.jl]: This comment belongs inside the function.

File: src/JuMP.jl [hunk 4]
```diff
@@ -350,6 +354,39 @@ bridge_constraints(model::Model) = moi_bridge_constraints(backend(model))
...
+    moi_add_bridge(caching_opt.optimizer, BridgeType)
+end
+
+"""
+     add_bridge(model::Model,
+                BridgeType::Type{<:MOI.Bridges.AbstractBridge})
+
+Add `BridgeType` to the list of bridges that can be used to transform
+unsupported constraints into an equivalent formulation using only constraints
+supported by the optimizer.
```
Comment (mlubin) [inline src/JuMP.jl]: Link to more documentation on bridges?

File: test/model.jl [hunk 5]
```diff
@@ -38,6 +38,9 @@ function opt_build(a::Int; b::Int = 1)
     return Optimizer(a, b)
 end
 
+# Custom set Nonnegative with bridge NonnegativeBridge
+include("nonnegative.jl")
```
Comment (mlubin) [inline test/model.jl]: `nonnegative_bridge.jl` is more informative

File: test/nonnegative.jl [hunk 6]
```diff
@@ -0,0 +1,64 @@
+using MathOptInterface
```
Comment (mlubin) [inline test/nonnegative.jl]: Missing license header.

File: test/nonnegative.jl [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+const MOI = MathOptInterface
+const MOIB = MOI.Bridges
+
```
Comment (mlubin) [inline test/nonnegative.jl]: It wouldn't hurt to add a comment at the top of the file explaining that this is an example bridge used for tests.

File: test/model.jl [hunk 8]
```diff
@@ -106,6 +109,70 @@ function test_model()
...
+            factory = with_optimizer(mock)
+            @testset "before loading the constraint to the optimizer" begin
+                @testset "with_optimizer at Model" begin
+                    model = Model(factory)
+                    @variable(model, x)
+                    JuMP.add_bridge(model, NonnegativeBridge)
+                    c = @constraint(model, x in Nonnegative())
+                    JuMP.optimize!(model)
+                    @test JuMP.value(x) == 1.0
+                    #@test JuMP.value(c) == 1.0
```
Comment (mlubin) [inline test/model.jl]: Why commented? (Same below)

File: test/model.jl [hunk 9]
```diff
@@ -106,6 +109,70 @@ function test_model()
...
+            @testset "after loading the constraint to the optimizer" begin
+                @testset "with_optimizer at Model" begin
+                    err = ErrorException(string("Constraints of type ",
+                    "MathOptInterface.SingleVariable-in-Nonnegative are not ",
+                    "supported by the solver and there are no bridges that ",
+                    "can reformulate it into supported constraints."))
+                    model = Model(factory)
+                    @variable(model, x)
+                    @test_throws err @constraint(model, x in Nonnegative())
+                    JuMP.add_bridge(model, NonnegativeBridge)
```
Comment (mlubin) [inline test/model.jl]: This is pretty cool.

File: src/JuMP.jl (hunk 4 above)
Comment (blegat) [reply src/JuMP.jl]: I have added it in the doc

File: test/model.jl (hunk 8 above)
Comment (blegat) [reply test/model.jl]: I was waiting for https://github.com/JuliaOpt/JuMP.jl/pull/1728 to be merged

File: src/JuMP.jl [hunk 10]
```diff
@@ -296,9 +300,11 @@ end
 
 Return mode (DIRECT, AUTOMATIC, MANUAL) of model.
 """
-mode(model::Model) = moi_mode(backend(model))
-# The type of backend(model) is unknown so we directly redirect to another
-# function.
+function mode(model::Model)
+    # The type of `backend(model)` is unknown so we directly redirect to another
+    # function.
+    moi_mode(backend(model))
```
Comment (odow) [inline src/JuMP.jl]: Nit: missing `return`

File: src/JuMP.jl [hunk 11]
```diff
@@ -296,9 +300,11 @@ end
 
 Return mode (DIRECT, AUTOMATIC, MANUAL) of model.
 """
-mode(model::Model) = moi_mode(backend(model))
-# The type of backend(model) is unknown so we directly redirect to another
-# function.
+function mode(model::Model)
+    # The type of `backend(model)` is unknown so we directly redirect to another
```
Comment (odow) [inline src/JuMP.jl]: Nit nit: Even though the comment was already there, `directly redirect` sounds weird. `so we redirect to` is fine.

File: src/JuMP.jl [hunk 12]
```diff
@@ -346,10 +352,45 @@ optimizer is set and unsupported constraints are automatically bridged
 to equivalent supported constraints when an appropriate transformation is
 available.
 """
-bridge_constraints(model::Model) = moi_bridge_constraints(backend(model))
-# The type of backend(model) is unknown so we directly redirect to another
-# function.
+function bridge_constraints(model::Model)
+    # The type of `backend(model)` is unknown so we directly redirect to another
```
Comment (odow) [inline src/JuMP.jl]: Ditto.

File: src/JuMP.jl [hunk 13]
```diff
@@ -346,10 +352,45 @@ optimizer is set and unsupported constraints are automatically bridged
...
+    # function.
+    return moi_bridge_constraints(backend(model))
+end
 
+function moi_add_bridge(model::Nothing,
+                        BridgeType::Type{<:MOI.Bridges.AbstractBridge})
+    # No optimizer is attached, the bridge will be added when one is attached
```
Comment (odow) [inline src/JuMP.jl]: Missing return

File: src/JuMP.jl [hunk 14]
```diff
@@ -346,10 +352,45 @@ optimizer is set and unsupported constraints are automatically bridged
...
+end
+function moi_add_bridge(model::MOI.ModelLike,
+                        BridgeType::Type{<:MOI.Bridges.AbstractBridge})
+    error("Cannot add bridge if `bridge_constraints` was set to `false` in the",
+          " `Model` constructor.")
+end
+function moi_add_bridge(bridge_opt::MOI.Bridges.LazyBridgeOptimizer,
+                        BridgeType::Type{<:MOI.Bridges.AbstractBridge})
+    MOI.Bridges.add_bridge(bridge_opt, BridgeType{Float64})
+end
```
Comment (odow) [inline src/JuMP.jl]: Missing `return` on a new line (I think we want to return nothing?)

File: src/JuMP.jl [hunk 15]
```diff
@@ -346,10 +352,45 @@ optimizer is set and unsupported constraints are automatically bridged
...
+function moi_add_bridge(caching_opt::MOIU.CachingOptimizer,
+                        BridgeType::Type{<:MOI.Bridges.AbstractBridge})
+    moi_add_bridge(caching_opt.optimizer, BridgeType)
+end
```
Comment (odow) [inline src/JuMP.jl]: Ditto

File: src/JuMP.jl [hunk 16]
```diff
@@ -346,10 +352,45 @@ optimizer is set and unsupported constraints are automatically bridged
...
+unsupported constraints into an equivalent formulation using only constraints
+supported by the optimizer.
+"""
+function add_bridge(model::Model,
+                    BridgeType::Type{<:MOI.Bridges.AbstractBridge})
+    push!(model.bridge_types, BridgeType)
+    # The type of `backend(model)` is unknown so we directly redirect to another
+    # function.
+    moi_add_bridge(JuMP.backend(model), BridgeType)
+end
```
Comment (odow) [inline src/JuMP.jl]: Ditto

File: src/JuMP.jl [hunk 17]
```diff
@@ -346,10 +352,45 @@ optimizer is set and unsupported constraints are automatically bridged
...
+
+Add `BridgeType` to the list of bridges that can be used to transform
+unsupported constraints into an equivalent formulation using only constraints
+supported by the optimizer.
+"""
+function add_bridge(model::Model,
+                    BridgeType::Type{<:MOI.Bridges.AbstractBridge})
+    push!(model.bridge_types, BridgeType)
+    # The type of `backend(model)` is unknown so we directly redirect to another
+    # function.
```
Comment (odow) [inline src/JuMP.jl]: What about:

The type of `backend(model)` is not type-stable, so we use a function barrier (`moi_add_bridge`) to improve performance.

File: src/optimizer_interface.jl [hunk 18]
```diff
@@ -50,6 +50,9 @@ function set_optimizer(model::Model, optimizer_factory::OptimizerFactory;
             optimizer = MOIU.CachingOptimizer(universal_fallback, optimizer)
         end
         optimizer = MOI.Bridges.full_bridge_optimizer(optimizer, Float64)
+        for BridgeType in model.bridge_types
```
Comment (odow) [inline src/optimizer_interface.jl]: Local variables should be named `bridge_type`

File: src/JuMP.jl (hunk 14 above)
Comment (blegat) [reply src/JuMP.jl]: `MOI.Bridges.add_bridge` already returns `nothing`. Is a `return` also needed in that case ?

Comment (odow) [reply src/JuMP.jl]: I've started doing it for clarity.

**Pull request merged**